decode/decode.go

```diff
@@ -0,0 +1,194 @@
+package decode
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/Clever/syslogparser/rfc3164"
+)
+
+// reservedFields are automatically set during decoding.
+// no field written by a user (e.g. contained in the Kayvee JSON) should overwrite them.
+var reservedFields = []string{
+	"prefix",
+	"postfix",
+	"Type",
+}
+
+func stringInSlice(s string, slice []string) bool {
+	for _, item := range slice {
+		if s == item {
+			return true
+		}
+	}
+	return false
+}
+
+// remapSyslog3164Keys renames fields to match our expecations from heka's syslog parser
+// see: https://github.com/mozilla-services/heka/blob/278dd3d5961b9b6e47bb7a912b63ce3faaf8d8bd/sandbox/lua/decoders/rsyslog.lua
+var remapSyslog3164Keys = map[string]string{
+	"hostname":  "Hostname",
+	"timestamp": "Timestamp",
+	"tag":       "programname",
+	"content":   "rawlog",
+}
+
+// FieldsFromSyslog takes an RSyslog formatted log line and extracts fields from it
+//
+// Supports two log lines formats:
+// - RSYSLOG_TraditionalFileFormat - the "old style" default log file format with low-precision timestamps (RFC3164)
+// - RSYSLOG_FileFormat - a modern-style logfile format similar to TraditionalFileFormat, but with high-precision timestamps and timezone information
+//
+// For more details on Rsylog formats: https://rsyslog-5-8-6-doc.neocities.org/rsyslog_conf_templates.html
+func FieldsFromSyslog(line string) (map[string]interface{}, error) {
+	// rfc3164 includes a severity number in front of the Syslog line, but we don't use that
+	fakeSeverity := "<12>"
+	p3164 := rfc3164.NewParser([]byte(fakeSeverity + line))
+	err := p3164.Parse()
+	if err != nil {
+		return map[string]interface{}{}, err
+	}
+
+	out := map[string]interface{}{}
+	for k, v := range p3164.Dump() {
+		if newKey, ok := remapSyslog3164Keys[k]; ok {
+			out[newKey] = v
+		}
+	}
+	return out, nil
+}
+
+type NonKayveeError struct{}
+
+func (e NonKayveeError) Error() string {
+	return fmt.Sprint("Log line is not Kayvee (doesn't have JSON payload)")
+}
+
+// FieldsFromKayvee takes a log line and extracts fields from the Kayvee (JSON) part
+func FieldsFromKayvee(line string) (map[string]interface{}, error) {
+	m := map[string]interface{}{}
+
+	firstIdx := strings.Index(line, "{")
+	lastIdx := strings.LastIndex(line, "}")
+	if firstIdx == -1 || lastIdx == -1 || firstIdx > lastIdx {
+		return map[string]interface{}{}, &NonKayveeError{}
+	}
+	m["prefix"] = line[:firstIdx]
+	m["postfix"] = line[lastIdx+1:]
+
+	possibleJSON := line[firstIdx : lastIdx+1]
+	var fields map[string]interface{}
+	if err := json.Unmarshal([]byte(possibleJSON), &fields); err != nil {
+		return map[string]interface{}{}, err
+	}
+	for k, v := range fields {
+		if !stringInSlice(k, reservedFields) {
+			m[k] = v
+		}
+	}
+
+	m["Type"] = "Kayvee"
+
+	return m, nil
```

Comment:
Since we were just using a JSON decoder before and calling it Kayvee, I think it's reasonable to not yet make it strict.

https://github.com/Clever/heka-clever-plugins/blob/master/lua/decoders/json_decoder.lua

If we need to distinguish between Kayvee and JSON, we can add that logic later